In [15]:
import pandas as pd  
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns




## Introductory Data Analysis

In [24]:
df = pd.read_csv("/Users/bera/Desktop/data/hotel_bookings.csv")

df.head(30)



hotel  is_canceled  lead_time  arrival_date_year  \
0   Resort Hotel            0        342               2015   
1   Resort Hotel            0        737               2015   
2   Resort Hotel            0          7               2015   
3   Resort Hotel            0         13               2015   
4   Resort Hotel            0         14               2015   
5   Resort Hotel            0         14               2015   
6   Resort Hotel            0          0               2015   
7   Resort Hotel            0          9               2015   
8   Resort Hotel            1         85               2015   
9   Resort Hotel            1         75               2015   
10  Resort Hotel            1         23               2015   
11  Resort Hotel            0         35               2015   
12  Resort Hotel            0         68               2015   
13  Resort Hotel            0         18               2015   
14  Resort Hotel            0         37               2015   
15  Resort Hotel            0         68               2015   
16  Resort Hotel            0         37               2015   
17  Resort Hotel            0         12               2015   
18  Resort Hotel            0          0               2015   
19  Resort Hotel            0          7               2015   
20  Resort Hotel            0         37               2015   
21  Resort Hotel            0         72               2015   
22  Resort Hotel            0         72               2015   
23  Resort Hotel            0         72               2015   
24  Resort Hotel            0        127               2015   
25  Resort Hotel            0         78               2015   
26  Resort Hotel            0         48               2015   
27  Resort Hotel            1         60               2015   
28  Resort Hotel            0         77               2015   
29  Resort Hotel            0         99               2015   

   arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0                July                        27                          1   
1                July                        27                          1   
2                July                        27                          1   
3                July                        27                          1   
4                July                        27                          1   
5                July                        27                          1   
6                July                        27                          1   
7                July                        27                          1   
8                July                        27                          1   
9                July                        27                          1   
10               July                        27                          1   
11               July                        27                          1   
12               July                        27                          1   
13               July                        27                          1   
14               July                        27                          1   
15               July                        27                          1   
16               July                        27                          1   
17               July                        27                          1   
18               July                        27                          1   
19               July                        27                          1   
20               July                        27                          1   
21               July                        27                          1   
22               July                        27                          1   
23               July                        27                          1   
24               July                        27                          1   
25               July        

In [17]:
print(df.shape)

(119390, 32)


In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [22]:
df.isna().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [19]:
# Understanding missingness in each column. 

missing_percentage = (df.isna().sum() / len(df)) * 100

columns_with_missing = missing_percentage[missing_percentage > 0]

print("Columns with missing values (percentage):")
print(columns_with_missing)


Columns with missing values (percentage):
children     0.003350
country      0.408744
agent       13.686238
company     94.306893
dtype: float64


### There are 4 columns with missing values.
- The agent column has 16340 (13.69%) missing values.
- The company column has 112593 (94.31%)missing values.
- The country column has 488 (0.41%)missing values.
- The children column has 4 (0.0034%) missing values.



In [23]:
# Unique Values in each column 
def unique_columns(df):
    for col in df.columns:
        uniques = df[col].unique()
        print(f"Number of Unique values: {len(uniques)}")
        print(f"Unique Values: {col}\n:{uniques}\n")

unique_columns(df)

Number of Unique values: 2
Unique Values: hotel
:['Resort Hotel' 'City Hotel']

Number of Unique values: 2
Unique Values: is_canceled
:[0 1]

Number of Unique values: 479
Unique Values: lead_time
:[342 737   7  13  14   0   9  85  75  23  35  68  18  37  12  72 127  78
  48  60  77  99 118  95  96  69  45  40  15  36  43  70  16 107  47 113
  90  50  93  76   3   1  10   5  17  51  71  63  62 101   2  81 368 364
 324  79  21 109 102   4  98  92  26  73 115  86  52  29  30  33  32   8
 100  44  80  97  64  39  34  27  82  94 110 111  84  66 104  28 258 112
  65  67  55  88  54 292  83 105 280 394  24 103 366 249  22  91  11 108
 106  31  87  41 304 117  59  53  58 116  42 321  38  56  49 317   6  57
  19  25 315 123  46  89  61 312 299 130  74 298 119  20 286 136 129 124
 327 131 460 140 114 139 122 137 126 120 128 135 150 143 151 132 125 157
 147 138 156 164 346 159 160 161 333 381 149 154 297 163 314 155 323 340
 356 142 328 144 336 248 302 175 344 382 146 170 166 338 167 310 148 165
